pallets/pools/src/lib.rs

```diff
@@ -628,6 +698,41 @@ pub mod pallet {
 				None => None,
 			};
 
+			for (tranche, tranche_input) in tranches.tranches.iter().zip(&tranche_inputs) {
+				let token_name: BoundedVec<u8, T::MaxTokenNameLength> = tranche_input
+					.clone()
+					.metadata
+					.token_name
+					.clone()
+					.try_into()
+					.map_err(|_| Error::<T>::TrancheTokenNameTooLong)?;
+
+				let token_symbol: BoundedVec<u8, T::MaxTokenSymbolLength> = tranche_input
+					.metadata
+					.token_symbol
+					.clone()
+					.try_into()
+					.map_err(|_| Error::<T>::TrancheSymbolNameTooLong)?;
+
+				let decimals = match T::AssetRegistry::metadata(&currency) {
+					Some(metadata) => metadata.decimals,
+					None => return Err(Error::<T>::MetadataForCurrencyNotFound.into()),
+				};
+
+				let parachain_id = T::ParachainId::get();
+
+				let metadata = create_asset_metadata(
+					decimals,
+					tranche.currency,
+					parachain_id,
+					token_name.to_vec(),
+					token_symbol.to_vec(),
+				);
+
+				T::AssetRegistry::register_asset(Some(tranche.currency), metadata)
+					.map_err(|_| Error::<T>::FailedToRegisterTrancheMetadata)?;
+			}
+
```

Comment:
It's not really possible to reduce this completely to a generic logic as we need `AssetRegistry` here. 

You could of course write something like `fn register_asset<Registry: orml_traits::registry::AssetRegistry()` but this feels a like it should be part of the pallet. 

What I think we can move to common is the boundaries check for vectors. With something like `fn respect_bounds<T, Bound: Get<u32>, E: Into<DispatchError>(vectors: Vec<T>) -> Result<BoundedVec<T, Bound>, DispatchError> {...}`. This would allow you to do the checks without the need to bind to `BoundedVec<T>` explicitly. 

Regarding the implementation. I think it would make sense to define a trait so that pallets can register pool info. We will most likely need this for the index-token too. And its light-coupling